File: samples/bot-adaptivecards-user-specific-views/csharp/Microsoft.Teams.Samples.UserSpecificViews/Cards/CardFactory.cs
// <copyright file="CardFactory.cs" company="Microsoft Corporation">
// Copyright (c) Microsoft Corporation. All rights reserved.
// </copyright>

namespace Microsoft.Teams.Samples.UserSpecificViews.Cards
{
    using System;
    using System.IO;
    using AdaptiveCards;
    using AdaptiveCards.Templating;
    using Microsoft.Bot.Schema;
    using Microsoft.Teams.Samples.UserSpecificViews.Bot;
    using Newtonsoft.Json;

    /// <summary>
    /// Card factory implementation.
    /// </summary>
    public class CardFactory : ICardFactory
    {
        private const string SelectCardTypeCardTemplatePath = "{0}\\assets\\templates\\select-card-type.json";
        private const string RefreshSpecificUserViewCardTemplatePath = "{0}\\assets\\templates\\refresh-specific-user.json";
        private const string RefreshAllUsersViewCardTemplatePath = "{0}\\assets\\templates\\refresh-all-users.json";
        private const string UpdatedBaseCardTemplatePath = "{0}\\assets\\templates\\updated-base-card.json";

        private const string BaseCardStatus = "Base";
        private const string UpdatedCardStatus = "Updated";
        private const string FinalCardStatus = "Final";

        private const string PersonalView = "Personal";
        private const string SharedView = "Shared";

        /// <summary>
        /// Initializes a new instance of the <see cref="CardFactory"/> class.
        /// </summary>
        /// <param name="appSettings">App settings.</param>
        public CardFactory()
        {
        }

        /// <inheritdoc/>
        public Attachment GetSelectCardTypeCard()
        {
            var data = new { };
            var template = GetCardTemplate(SelectCardTypeCardTemplatePath);
            var serializedJson = template.Expand(data);
            return CreateAttachment(serializedJson);
        }

        public Attachment GetAutoRefreshForAllUsersBaseCard(string cardType)
        {
            var count = 0; // Initial count is 0 for base card.
            var data = new
            {
                count,
                cardType,
                cardStatus = BaseCardStatus,
                trigger = "NA",
                view = SharedView,
                message = "Original Message"
            };

            var template = GetCardTemplate(RefreshAllUsersViewCardTemplatePath);
            var serializedJson = template.Expand(data);
            return CreateAttachment(serializedJson);
        }

        /// <inheritdoc/>
        public Attachment GetAutoRefreshForSpecificUserBaseCard(string userMri, string cardType)
        {
            var count = 0;
            var data = new
            {
                count,
                cardType,
                cardStatus = BaseCardStatus,
                trigger = "NA",
                view = SharedView,
                userMri,
                message = "Original Message"
            };

            var template = GetCardTemplate(RefreshSpecificUserViewCardTemplatePath);
            var serializedJson = template.Expand(data);
            return CreateAttachment(serializedJson);
        }

        /// <inheritdoc/>
        public Attachment GetUpdatedCardForUser(string userMri, RefreshActionData actionData)
        {
            var data = new
            {
                count = actionData.action.data.RefreshCount,
                userMri = userMri,
                cardType = actionData.action.data.CardType,
                cardStatus = UpdatedCardStatus,
                trigger = actionData.trigger,
                view = PersonalView,
                message = "Updated Message!"
            };

            var template = GetCardTemplate(RefreshSpecificUserViewCardTemplatePath);
            var serializedJson = template.Expand(data);
            return CreateAttachment(serializedJson);
        }

        

        /// <inheritdoc/>
        public Attachment GetFinalBaseCard(RefreshActionData actionData)
        {
            var data = new
            {
                count = actionData.action.data.RefreshCount,
                cardType = actionData.action.data.CardType,
                cardStatus = FinalCardStatus,
                trigger = actionData.trigger,
                view = SharedView,
                message = "Final Message!"
            };

            var template = GetCardTemplate(UpdatedBaseCardTemplatePath);
            var serializedJson = template.Expand(data);
            return CreateAttachment(serializedJson);
        }

        private static AdaptiveCardTemplate GetCardTemplate(string templatePath)
        {
            templatePath = string.Format(templatePath, AppDomain.CurrentDomain.BaseDirectory);
            return new AdaptiveCardTemplate(File.ReadAllText(templatePath));
        }

        private static Attachment CreateAttachment(string adaptiveCardJson)
        {
            var adaptiveCard = AdaptiveCard.FromJson(adaptiveCardJson);
            return new Attachment
            {
                ContentType = AdaptiveCard.ContentType,
                Content = JsonConvert.DeserializeObject(adaptiveCardJson),
            };
        }
    }
}

File: samples/bot-adaptivecards-user-specific-views/python/card_factory/card_factory.py
import os
import json
from jinja2 import Template
from botbuilder.schema import Attachment
from typing import Any

BASE_DIR = os.path.dirname(os.path.abspath(__file__))

class CardFactory:
    SELECT_CARD_TYPE_TEMPLATE = os.path.join(BASE_DIR, "..", "templates", "select-card-type.json")
    REFRESH_SPECIFIC_USER_TEMPLATE = os.path.join(BASE_DIR, "..", "templates", "refresh-specific-user.json")
    REFRESH_ALL_USERS_TEMPLATE = os.path.join(BASE_DIR, "..", "templates", "refresh-all-users.json")
    UPDATED_BASE_CARD_TEMPLATE = os.path.join(BASE_DIR, "..", "templates", "updated-base-card.json")

    BASE_CARD_STATUS = "Base"
    UPDATED_CARD_STATUS = "Updated"
    FINAL_CARD_STATUS = "Final"

    PERSONAL_VIEW = "Personal"
    SHARED_VIEW = "Shared"

    def get_select_card_type_card(self) -> Attachment:
        data = {}
        template = self._get_card_template(self.SELECT_CARD_TYPE_TEMPLATE)
        card_json = template.render(**data)
        return self._create_attachment(card_json)

    def get_auto_refresh_for_all_users_base_card(self, card_type: str) -> Attachment:
        data = {
            "count": 0,
            "cardType": card_type,
            "cardStatus": self.BASE_CARD_STATUS,
            "trigger": "NA",
            "view": self.SHARED_VIEW,
            "message": "Original Message"
        }
        template = self._get_card_template(self.REFRESH_ALL_USERS_TEMPLATE)
        card_json = template.render(**data)
        return self._create_attachment(card_json)

    def get_auto_refresh_for_specific_user_base_card(self, user_mri: str, card_type: str) -> Attachment:
        data = {
            "count": 0,
            "cardType": card_type,
            "cardStatus": self.BASE_CARD_STATUS,
            "trigger": "NA",
            "view": self.SHARED_VIEW,
            "userMri": user_mri,
            "message": "Original Message"
        }
        template = self._get_card_template(self.REFRESH_SPECIFIC_USER_TEMPLATE)
        card_json = template.render(**data)
        return self._create_attachment(card_json)

    def get_updated_card_for_user(self, user_mri: str, action_data: Any) -> Attachment:
        data = {
            "count": action_data["action"]["data"]["refreshCount"],
            "cardType": action_data["action"]["data"]["cardType"],
            "cardStatus": self.UPDATED_CARD_STATUS,
            "trigger": action_data["trigger"],
            "view": self.PERSONAL_VIEW,
            "userMri": user_mri,
            "message": "Updated Message!"
        }
        template = self._get_card_template(self.REFRESH_SPECIFIC_USER_TEMPLATE)
        card_json = template.render(**data)
        return self._create_attachment(card_json)

    def get_final_base_card(self, action_data: Any) -> Attachment:
        data = {
            "count": action_data["action"]["data"]["refreshCount"],
            "cardType": action_data["action"]["data"]["cardType"],
            "cardStatus": self.FINAL_CARD_STATUS,
            "trigger": action_data["trigger"],
            "view": self.SHARED_VIEW,
            "message": "Final Message!"
        }
        template = self._get_card_template(self.UPDATED_BASE_CARD_TEMPLATE)
        card_json = template.render(**data)
        return self._create_attachment(card_json)

    def _get_card_template(self, path: str) -> Template:
        with open(path, encoding="utf-8") as f:
            content = f.read()
            return Template(content)

    def _create_attachment(self, card_json: str) -> Attachment:
        content = json.loads(card_json)
        return Attachment(content_type="application/vnd.microsoft.card.adaptive", content=content)